### PR TITLE
authz in resolver

### PR DESCRIPTION
**Pull request by tim-harper**

### CODE REVIEW

File: internal/auth/auth.go [hunk 1]
```diff
@@ -26,7 +26,9 @@ func WithAuthedUser(ctx context.Context, user *DelphisAuthedUser) context.Contex
 
 func GetAuthedUser(ctx context.Context) *DelphisAuthedUser {
 	val := ctx.Value(authedUserContextString)
+	logrus.Debugf("Val: %v\n", val)
 	if valAsAuthedUser, ok := val.(*DelphisAuthedUser); ok {
+		logrus.Debugf("Does this ever return the User or just ID: %v   %+v\n", valAsAuthedUser.UserID, valAsAuthedUser.User)
```
Comment (tim-harper) [inline internal/auth/auth.go:31]: ??

File: graph/resolver/schema.resolvers.go [hunk 2]
```diff
@@ -96,17 +96,26 @@ func (r *mutationResolver) CreateDiscussion(ctx context.Context, anonymityType m
 }
 
 func (r *mutationResolver) CreateFlair(ctx context.Context, userID string, templateID string) (*model.Flair, error) {
+	if userID == "" || templateID == "" {
```
Comment (tim-harper) [inline graph/resolver/schema.resolvers.go:99]: add checks to all calls

File: graph/resolver/schema.resolvers.go [hunk 3]
```diff
@@ -13,10 +13,13 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-func (r *mutationResolver) AddDiscussionParticipant(ctx context.Context, discussionID string, userID string, discussionParticipantInput model.AddDiscussionParticipantInput) (*model.Participant, error) {
-	participantObj, err := r.DAOManager.CreateParticipantForDiscussion(ctx, discussionID, userID, discussionParticipantInput)
+func (r *mutationResolver) AddDiscussionParticipant(ctx context.Context, discussionID string, discussionParticipantInput model.AddDiscussionParticipantInput) (*model.Participant, error) {
+	authedUser := auth.GetAuthedUser(ctx)
```
Comment (tim-harper) [inline graph/resolver/schema.resolvers.go:17]: align on `authedUser`

File: graph/resolver/schema.resolvers.go [hunk 4]
```diff
@@ -270,8 +304,24 @@ func (r *mutationResolver) UpdateParticipant(ctx context.Context, discussionID s
...
 func (r *mutationResolver) UpsertUserDevice(ctx context.Context, userID *string, platform model.Platform, deviceID string, token *string) (*model.UserDevice, error) {
-	// TODO: Check authz
-	resp, err := r.DAOManager.UpsertUserDevice(ctx, deviceID, userID, platform.String(), token)
+	currentUser := auth.GetAuthedUser(ctx)
+	if currentUser == nil {
+		return nil, fmt.Errorf("Need auth")
+	}
+
+	// Check if the user already has a device registered
+	userDevice, err := r.DAOManager.GetUserDeviceByUserIDPlatform(ctx, *userID, platform.String())
```
Comment (tim-harper) [inline graph/resolver/schema.resolvers.go:313]: should we support a user having multiple devices on the same platform? probably.

File: graph/resolver/schema.resolvers.go (hunk 3 above)
Comment (nedrocks) [reply graph/resolver/schema.resolvers.go:17]: Do you think we should check whether the user is the moderator or sth?

File: graph/resolver/schema.resolvers.go [hunk 5]
```diff
@@ -145,12 +154,17 @@ func (r *mutationResolver) RemoveFlair(ctx context.Context, id string) (*model.F
...
-	// behalf of another user. To add this functionality remove the following
-	// line and add authorization/permission logic.
+	// An admin or moderator could potentially remove flair on
+	// behalf of another user.
 	if currentUser.UserID != flair.UserID {
-		return nil, fmt.Errorf("Unauthorized")
+		// Check if a moderator is trying to removeFlair
+		mod, err := r.DAOManager.GetModeratorByUserID(ctx, currentUser.UserID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch moderator")
```
Comment (nedrocks) [inline graph/resolver/schema.resolvers.go:163]: I think we should just return unauthorized in these cases -- we wouldn't want to expose this to the client I think

File: graph/resolver/schema.resolvers.go (hunk 4 above)
Comment (nedrocks) [reply graph/resolver/schema.resolvers.go:313]: Yeah good point!

File: graph/types/schema.graphqls [hunk 6]
```diff
@@ -51,7 +51,7 @@ input PostContentInput {
 
 
 type Mutation {
-  addDiscussionParticipant(discussionID: String!, userID: String!, discussionParticipantInput: AddDiscussionParticipantInput!): Participant!
+  addDiscussionParticipant(discussionID: String!, discussionParticipantInput: AddDiscussionParticipantInput!): Participant!
```
Comment (nedrocks) [inline graph/types/schema.graphqls:54]: So this endpoint, I think, needs to have userID so the moderator could add a user.

File: internal/auth/auth.go (hunk 1 above)
Comment (nedrocks) [reply internal/auth/auth.go:31]: what's this one? Sorry I'm not following!

**Pull request merged**